hsmtools/repack.man
.\" ******************************************************************************
.\"                      repack
.\"
.\" This file is part of the Castor project.
.\" See http://castor.web.cern.ch/castor
.\"
.\" Copyright (C) 2003  CERN
.\" This program is free software; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License
.\" as published by the Free Software Foundation; either version 2
.\" of the License, or (at your option) any later version.
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
.\"
.\" man page for the repack command.
.\"
.\" @author Castor Dev team, castor-dev@cern.ch
.\" *****************************************************************************/
.TH repack "8castor" "July, 2011" CASTOR "Moves data away from a tape"
.SH NAME
repack \- Move files from tape(s) to other tape destination
.SH SYNOPSIS
.B repack
.BI -h
.br
.B repack
.BI \-V\ <VID>[:<VID>[...]]\ \-o\ OutputServiceClass
.br
.B repack
.BI \-\-bulkvolumeid\ <file>\ \-o\ OutputServiceClass
.br
.B repack
.BI \-R\ <VID>[:<VID>[...]]
.br
.B repack
.BI \-\-bulkdelete\ <file>
.br
.B repack
.BI \-s|--statusAll\ [-x|--extra]
.br
.B repack
.BI \-S\ <VID>[:<VID>[...]]\ [-x|--extra]
.br
.B repack
.BI \-e\ <VID>[:<VID>[...]]\ [\-n\ <nbErrorsDisplayed>]
.br

.SH DESCRIPTION
.B repack
allows to move files from one or several tapes to new media.
It also allows to get the status overview of the ongoing repack activity in the system.
Note that only volumes having
.B FULL
status may be repacked.

When the repack process has been successfully completed (FINISHED status), the volume can be
reclaimed using the reclaim command.

.SH OPTIONS

.TP
.BI \-h,\ \-\-help
displays command usage help.
.TP
.BI \-V,\ \-\-volumeid
submits the repacking of the given tapes and returns immediately. The tapes are given by their VID
and the list is colon separated. The actual repack processes are asynchronously started, one by one,
by the database.
.TP
.BI \-\-bulkvolumeid
submits the repacking of the tapes given in the provided file. The file must contain a list of VIDs,
one per line. Spaces and anything appearing after the '#' sign on a given line are ignored.
.TP
.BI \-o,\ \-\-svcclass
the service class to be used for repacking the given tapes.
.TP
.BI \-R,\ \-\-delete
cancels the repacking of the given tapes. The tapes are given by their VID and the list is colon separated.
Note that the cancelation may be refused in case too many repacks are already being launched or aborted.
The maximum number is defined in the Repack/MaxNbConcurrentClients entry of the CastorConfig table
in the stager database. The default value is 5.
.TP
.BI \-\-bulkdelete
cancels the repacking of the tapes given in the provided file. The file must contain alist of VIDs,
one per line. Spaces and anything appearing after the '#' sign on a given line are ignored.
Note that the cancelation may be refused in case too many repacks are already being launched or aborted.
The maximum number is defined in the Repack/MaxNbConcurrentClients entry of the CastorConfig table
in the stager database. The default value is 5.
.TP
.BI \-s,\ \-\-statusAll
shows the status of all ongoing and finished repacks. If multiple repack requests were issued for a given
tape, only the last request is shown. Note that old completed repacks are automatically removed
from the database and are not shown any longer.
.TP
.BI \-S,\ \-\-status
shows the status of ongoing and finished repacks for the given tape(s). The tape(s) is(are) given by their
VID and the list is colon separated. As opposed to the \-s option, here the entire history is shown.
.TP
.BI \-x,\ \-\-extra
shows the status of ongoing and finished repacks as requested with the -s or -S options, including extra
detailed figures for the count of ongoing or failed recalls and migrations, and a completion percentage.
.TP
.BI \-e,\ \-\-errors
displays the errors that occured during the repack of a given tape, file by file. By default, the number of
errors displayed is limited to 10. This can be changed using the \-n option
.TP
.BI \-n,\ \-\-nbErrorsDisplayed
in conjunction with \-e, sets the maximum number of errors to display.

.SH EXAMPLE
.nf
.ft CW
> repack -V V00000:V10025:V13049 -o default
vmgrlisttape: No such tape
Caught exception when executing vmgrlisttape for tape V00000, skipping
Tape V10025 is not marked FULL, ignoring it
Tape V13049 is already being repacked by user root, ignoring it

> repack -V I10551:I10552 -o default
Repack submitted for tape I10551
Repack submitted for tape I10552

> repack -s
=====================================================================================================================
SubmitTime        RepackTime             User                      Machine      Vid      Total        Size     Status
---------------------------------------------------------------------------------------------------------------------
11-Jun-13 18:57       1mn20s    itglp@CERN.CH             lxc2dev2.cern.ch   V21002          4     3.29KiB   FINISHED
10-Jun-13 15:42       1mn40s             root             lxc2dev2.cern.ch   V22003          1     2.93KiB   FINISHED
---------------------------------------------------------------------------------------------------------------------
11-Jun-13 19:29            -                -                        TOTAL        2          5     6.22KiB   FINISHED

> repack -sx
=====================================================================================================================================================================
SubmitTime        RepackTime             User                      Machine      Vid      Total        Size  toRecall    toMigr    Failed  Migrated  Compl%     Status
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
11-Jun-13 18:57       1mn20s    itglp@CERN.CH             lxc2dev2.cern.ch   V21002          4     3.29KiB         0         0         0         4    100%   FINISHED
10-Jun-13 15:42       1mn40s             root             lxc2dev2.cern.ch   V22003          1     2.93KiB         0         0         0         1    100%   FINISHED
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
11-Jun-13 19:29            -                -                        TOTAL        2          5     6.22KiB         0         0         0         5    100%   FINISHED

> repack -S V21002 -x

hsmtools/repack.c
/*
 * Copyright (C) 2001-2003 by CERN/IT/PDP/DM
 * All rights reserved
 */

/*      repack - copy the active segments from a set of volumes to another set */
#include <errno.h>
#include <pwd.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <time.h>
#include <stdarg.h>
#if defined(_WIN32)
#include <winsock2.h>
#endif
#include "Cgetopt.h"
#include "Cns_api.h"
#include "Ctape_api.h"
#include "log.h"
#include "rtcp_api.h"
#include "serrno.h"
#include "stage_api.h"
#include "u64subr.h"
#include "vmgr_api.h"
struct vol_elem {
	char vid[CA_MAXVIDLEN+1];
	char poolname[CA_MAXPOOLNAMELEN+1];
};
void cleanup _PROTO((int));
int help_flag;
char *host;
int last_rtcp_disk_fseq;
int last_seg_repacked;
int last_stcp_repacked;
int nodelete_flag;
int nstcp_output;
int out_side;
char *out_vid;
int rtcpc_kill_needed;
struct Cns_direntape *seg_list;
int seg_list_size;
u_signed64 size_requested;
int stage_kill_needed;
struct stgcat_entry *stcp_output;
int sys_errflg;
char *timestamp();
int usr_errflg;
struct vol_elem *vol_list;
int vol_list_size;
int use_checksum;
extern int (*rtcpc_ClientCallback) _PROTO((rtcpTapeRequest_t *, rtcpFileRequest_t *));
EXTERN_C char DLL_DECL *getconfent _PROTO(());

#define FILL_USE_CHECKSUM_VARIABLE() {                \
	char *p;                                          \
	use_checksum = 1;                                 \
	if ((p = getenv("REPACK_USE_CHECKSUM")) == NULL) {\
		p = getconfent("REPACK","USE_CHECKSUM",0);    \
	}                                                 \
	if (p != NULL) {                                  \
		if (strcmp(p,"NO") == 0) {                    \
			use_checksum = 0;                         \
		}                                             \
	}                                                 \
}

main(argc, argv)
int argc;
char **argv;
{
	int c;
	char density[CA_MAXDENLEN+1];
	char *dp;
	int errflg = 0;
	int i;
	int iflag = 0;
	char library[CA_MAXTAPELIBLEN+1];
	char localhost[CA_MAXHOSTNAMELEN+1];
	static struct Coptions longopts[] = {
		{"help", NO_ARGUMENT, &help_flag, 1},
		{"library", REQUIRED_ARGUMENT, 0, OPT_LIBRARY_NAME},
		{"min_free", REQUIRED_ARGUMENT, 0, 'm'},
		{"model", REQUIRED_ARGUMENT, 0, OPT_MODEL},
		{"nodelete", NO_ARGUMENT, &nodelete_flag, 1},
		{"output_tppool", REQUIRED_ARGUMENT, 0, 'o'},
		{"otp", REQUIRED_ARGUMENT, 0, 'o'},
		{0, 0, 0, 0}
	};
	int max_vol_to_repack = 0;
	int min_free = 0;
	char model[CA_MAXMODELLEN+1];
	int nbvol;
	char output_pool[CA_MAXPOOLNAMELEN+1];
	struct passwd *pw;
	char pool_name[CA_MAXPOOLNAMELEN+1];
	char *vids = NULL;
#if defined(_WIN32)
	WSADATA wsadata;
#endif

	Copterr = 1;
	Coptind = 1;
	density[0] = '\0';
	library[0] = '\0';
	model[0] = '\0';
	output_pool[0]= '\0';
	pool_name[0]= '\0';
	while ((c = Cgetopt_long (argc, argv, "d:h:in:P:V:", longopts, NULL)) != EOF) {
		switch (c) {
		case OPT_LIBRARY_NAME:
			if (strlen (Coptarg) <= CA_MAXTAPELIBLEN)
				strcpy (library, Coptarg);
			else {
				fprintf (stderr,
				    "invalid library %s\n", Coptarg);
				errflg++;
			}
			break;
		case OPT_MODEL:
			if (strlen (Coptarg) <= CA_MAXMODELLEN)
				strcpy (model, Coptarg);
			else {
				fprintf (stderr, "%s: %s\n", Coptarg,
				    "model name too long");
				errflg++;
			}
			break;
		case 'd':
			if (strlen (Coptarg) <= CA_MAXDENLEN)
				strcpy (density, Coptarg);
			else {
				fprintf (stderr,
				    "invalid density %s\n", Coptarg);
				errflg++;
			}
			break;
		case 'h':
			host = Coptarg;
			break;
		case 'i':
			iflag++;
			break;
		case 'm':
			min_free = strtol (Coptarg, &dp, 10);
			if (*dp != '\0') {
				fprintf (stderr, "invalid min_free value: %s\n",
				    Coptarg);
				errflg++;
			}
		case 'n':
			max_vol_to_repack = strtol (Coptarg, &dp, 10);
			if (*dp != '\0' || max_vol_to_repack <= 0) {
				fprintf (stderr, "invalid number of volumes: %s\n",
				    Coptarg);
				errflg++;
			}
			break;
		case 'o':
			if (strlen (Coptarg) <= CA_MAXPOOLNAMELEN)
				strcpy (output_pool, Coptarg);
			else {
				fprintf (stderr, "%s\n", strerror(EINVAL));
				errflg++;
			}
			break;
		case 'P':
			if (strlen (Coptarg) <= CA_MAXPOOLNAMELEN)
				strcpy (pool_name, Coptarg);
			else {
				fprintf (stderr, "%s\n", strerror(EINVAL));
				errflg++;
			}
			break;
		case 'V':
			vids = Coptarg;
			break;
		case '?':
			errflg++;
			break;
		default:
			break;
		}
	}
	if (Coptind < argc || (! vids && *pool_name == 0 && iflag == 0)) {
		errflg++;
	}
	if (errflg || help_flag) {
		fprintf (stderr, "usage: %s %s%s%s", argv[0],
		    "[-d density] [-h name_server] [--help] [-i] [--li library_name]\n",
		    "\t[--min_free n] [--mo model] [-n max_vol_to_repack] [--nodelete]\n",
		    "\t[--otp output_tppool] [-P pool_name] [-V vid(s)]\n");
		exit (USERR);
	}
#if defined(_WIN32)
	if (WSAStartup (MAKEWORD (2, 0), &wsadata)) {
		fprintf (stderr, "WSAStartup unsuccessful\n");
		exit (SYERR);
	}
#endif
	gethostname (localhost, CA_MAXHOSTNAMELEN+1);
	pw = getpwuid (getuid());
	printf ("%s Repack run by %s on %s\n", timestamp(), pw->pw_name, localhost);
	fflush (stdout);

	if (iflag)
		nbvol = get_vidlist_from_stdin (pool_name);
	else if (vids)
		nbvol = get_vidlist_from_cmdline (vids, pool_name);
	else
		nbvol = get_vidlist_from_db (pool_name, library, model, density,
			max_vol_to_repack);
	if (nbvol == 0) {
#if defined(_WIN32)
		WSACleanup();
#endif
		exit (usr_errflg ? USERR : (sys_errflg ? SYERR : 0));
	}

#if ! defined(_WIN32)
	signal (SIGHUP, cleanup);
#endif
	signal (SIGINT, cleanup);
#if ! defined(_WIN32)
	signal (SIGQUIT, cleanup);
#endif
	signal (SIGTERM, cleanup);

	/* copy active segments */

	for (i = 0; i < nbvol; i++) {
		if (c = copy_active_segments (vol_list[i].vid, *output_pool ?
		    output_pool : vol_list[i].poolname)) break;
	}
#if defined(_WIN32)
	WSACleanup();
#endif
	exit (c);
}

addvid(volindex, vid, pool_name)
int volindex;
char *vid;
char *pool_name;
{
	if ((CA_MAXVIDLEN+1) * (volindex+1) > vol_list_size) {
		vol_list_size += BUFSIZ;
		vol_list = realloc (vol_list, vol_list_size);
	}
	strcpy (vol_list[volindex].vid, vid);
	strcpy (vol_list[volindex].poolname, pool_name);
	return (0);
}

checkvol(vid, pool_name)
char *vid;
char *pool_name;
{
	char p_stat[9];
	struct vmgr_tape_info tape_info;

	/* check if the volume exists and is marked FULL */

	if (vmgr_querytape (vid, 0, &tape_info, NULL) < 0) {
		fprintf (stderr, "repack %s: %s\n", vid,
		    (serrno == ENOENT) ?
		    "No such volume" : sstrerror(serrno));
		usr_errflg++;
		return (-1);
	}
	if ((tape_info.status & TAPE_FULL) == 0) {
		if (tape_info.status == 0) strcpy (p_stat, "FREE");
		else if (tape_info.status & TAPE_BUSY) strcpy (p_stat, "BUSY");
		else if (tape_info.status & TAPE_RDONLY) strcpy (p_stat, "RDONLY");
		else if (tape_info.status & EXPORTED) strcpy (p_stat, "EXPORTED");
		else if (tape_info.status & DISABLED) strcpy (p_stat, "DISABLED");
		else strcpy (p_stat, "?");
		fprintf (stderr, "Volume %s has status %s\n", vid, p_stat);
		usr_errflg++;
		return (-1);
	}
	strcpy (pool_name, tape_info.poolname);
	return (0);
}

void cleanup(sig)
int sig;
{
	signal (sig, SIG_IGN);
	if (rtcpc_kill_needed)
		rtcpc_kill();
	if (out_vid) {
		if (vmgr_updatetape (out_vid, out_side, 0, 0, 0, 0) < 0) {
			fprintf (stderr, "vmgr_updatetape %s: %s\n", out_vid,
			    sstrerror(serrno));
		}
	}
	if (stage_kill_needed)
		stage_kill (sig);
#if defined(_WIN32)
	WSACleanup();
#endif
	exit (USERR);
}

copy_active_segments(vid, pool_name)
char *vid;
char *pool_name;
{
	int c = 0;
	struct Cns_direntape *dtp;
	int flags;
	int i;
	int j;
	Cns_list list;
	int nbsegs = 0;
	struct stgcat_entry *stcp;
	struct stgcat_entry *stcp_input;

	printf ("%s Processing %s\n", timestamp(), vid);
	fflush (stdout);
	serrno = 0;
	flags = CNS_LIST_BEGIN;
	while ((dtp = Cns_listtape (host, vid, flags, &list, 0)) != NULL) {
		flags = CNS_LIST_CONTINUE;
		if (dtp->s_status == 'D') continue;
		if ((nbsegs+1) * sizeof(struct Cns_direntape) > seg_list_size) {
			seg_list_size += BUFSIZ;
			if ((seg_list = realloc (seg_list, seg_list_size)) == NULL) {
				fprintf (stderr, "repack error: %s\n", strerror(ENOMEM));
				return (USERR);
			}
		}
		seg_list[nbsegs] = *dtp;
		size_requested += dtp->segsize;
		nbsegs++;
	}
	if (serrno > 0 && serrno != ENOENT) {
		fprintf (stderr, "repack %s: %s\n", vid, sstrerror(serrno));
		return (USERR);
	}
	if (serrno == 0) {
		(void) Cns_listtape (host, vid, CNS_LIST_END, &list, 0);
	}

	printf ("%s Vid %s: %d segments to repack\n", timestamp(), vid, nbsegs);
	fflush (stdout);

	if ((stcp_input = calloc ((nbsegs <= stage_util_max_stcp_per_request()) ? nbsegs : stage_util_max_stcp_per_request(),
	    sizeof(struct stgcat_entry))) == NULL) {
		fprintf (stderr, "repack error: %s\n", strerror(ENOMEM));
		return (USERR);
	}
	j = 0;
	last_seg_repacked = 0;
	stcp = stcp_input;
	for (i = 0; i < nbsegs; i++) {
		strcpy (stcp->recfm, "F");
		stcp->size = seg_list[i].segsize + 1;
		sprintf (stcp->u1.t.fseq, "%d", seg_list[i].fseq);
		strcpy (stcp->u1.t.vid[0], vid);
		j++;
		stcp++;
		if (j == stage_util_max_stcp_per_request() || i + 1 == nbsegs) {
			printf ("%s Stageing in %s set of segments from %s (%d file sequence%s)\n",
			    timestamp(), i <= stage_util_max_stcp_per_request() ? "first" : (i + 1 == nbsegs) ?
			    "last" : "next", vid, j, (j > 1) ? "s" : "");
			fflush (stdout);
			stage_kill_needed = 1;
			if (stage_in_tape ((u_signed64) STAGE_DEFERRED,
			    O_RDONLY, NULL, NULL, j, stcp_input, &nstcp_output,
			    &stcp_output, 0, NULL) < 0) {
				fprintf (stderr, "stage_in_tape %s error: %s\n", vid,
						 sstrerror(serrno));
				c = rc_castor2shift(serrno);
				break;
			}
			stage_kill_needed = 0;
			last_stcp_repacked = 0;
			while (last_stcp_repacked < nstcp_output) {
				if (c = write_segments (pool_name)) break;
			}
			free (stcp_output);
			if (c) break;
			j = 0;
			stcp = stcp_input;
		}
	}
	free (seg_list);
	seg_list = NULL;
	seg_list_size = 0;
	free (stcp_input);
	if (c == 0)
		printf ("%s Vid %s successfully repacked\n", timestamp(), vid);
	else
		printf ("%s Repack of %s failed\n", timestamp(), vid);
	fflush (stdout);
	return (c);
}

get_vidlist_from_cmdline(vids, pool_name)
char *vids;
char *pool_name;
{
	int nbvol = 0;
	char *vid;

	for (vid = strtok (vids, ":"); vid;  vid = strtok (NULL, ":")) {
		if (checkvol (vid, pool_name) == 0) {
			addvid (nbvol, vid, pool_name);
			nbvol++;
		}
	}
	return (nbvol);
}

get_vidlist_from_db(pool_name, library, model, density, max_vol_to_repack)
char *pool_name;
char *library;
char *model;
char *density;
int max_vol_to_repack;
{
	int flags;
	vmgr_list list;
	struct vmgr_tape_info *lp;
	int nbvol = 0;

	/* check if pool_name exists */

	if (vmgr_querypool (pool_name, NULL, NULL, NULL, NULL) < 0) {
		fprintf (stderr, "repack %s: %s\n", pool_name,
		    (serrno == ENOENT) ?
		    "No such pool" : sstrerror(serrno));
		usr_errflg++;
		return (0);
	}

	/* get list of volumes with status FULL in pool_name */

	flags = VMGR_LIST_BEGIN;
	while ((lp = vmgr_listtape (NULL, pool_name, flags, &list)) != NULL) {
		flags = VMGR_LIST_CONTINUE;
		if ((lp->status & TAPE_FULL) == 0) continue;
		if (*library && strcmp (lp->library, library)) continue;
		if (*model && strcmp (lp->model, model)) continue;
		if (*density && strcmp (lp->density, density)) continue;
		addvid (nbvol, lp->vid, pool_name);
		nbvol++;
		if (max_vol_to_repack && nbvol >= max_vol_to_repack)
			break;
	}
	(void) vmgr_listtape (NULL, pool_name, VMGR_LIST_END, &list);

	if (nbvol == 0) {
		fprintf (stderr,
		    "No volume with status FULL in pool %s\n",
		    pool_name);
	}
	return (nbvol);
}

get_vidlist_from_stdin(pool_name)
char *pool_name;
{
	int n;
	int nbvol = 0;
	char vid[256];

	while (fgets (vid, sizeof(vid), stdin) != NULL) {
		n = strlen (vid) - 1;
		if (vid[n] != '\n' || n > CA_MAXVIDLEN) {
			fprintf (stderr, "Vid %s too long\n", vid);
			usr_errflg++;
		} else {
			vid[n] = '\0';
			if (checkvol (vid, pool_name) == 0) {
				addvid (nbvol, vid, pool_name);
				nbvol++;
			}
		}
	}
	return (nbvol);
}

repack_callback (rtcpTapeRequest_t *tl, rtcpFileRequest_t *fl)
{
	int compression_factor = 0;
	u_signed64 fileid;
	int Flags;
	struct Cns_segattrs new_segattrs;
	struct Cns_segattrs old_segattrs;
	char tmpbuf[21];

	if (fl->cprc == 0 && fl->proc_status == RTCP_FINISHED) {
		/* segment successfully copied to tape */
		Flags = (fl->disk_fseq == last_rtcp_disk_fseq) ? 0 : TAPE_BUSY;
		if (fl->bytes_out > 0)
			compression_factor = fl->host_bytes * 100 / fl->bytes_out;

		if (vmgr_updatetape (tl->vid, tl->side, fl->bytes_in,
		    compression_factor, 1, Flags) < 0) {
			fprintf (stderr, "vmgr_updatetape %s: %s\n", tl->vid,
			    sstrerror(serrno));
			if (serrno == EACCES)
				usr_errflg++;
			else
				sys_errflg++;
			return (-1);
		}

		FILL_USE_CHECKSUM_VARIABLE();
		if (use_checksum && (fl->castorSegAttr.segmCksumAlgorithm[0] != '\0')) {
			/* We verify from the tpread with the checksum we got from name server */
			if (strcmp(fl->castorSegAttr.segmCksumAlgorithm,seg_list[last_seg_repacked].checksum_name) == 0) {
				/* Same algorithm - then the values have to match */
				if (seg_list[last_seg_repacked].checksum != fl->castorSegAttr.segmCksum) {
					serrno = SECHECKSUM;
					fprintf (stderr, "%s (%s,%d,%d) repacked from (%s,%d,%d,%02x%02x%02x%02x,%d,\"%s\",0x%lx) to (%s,%d,%d,%02x%02x%02x%02x,%d, \"%s\", 0x%lx) error: %s\n",
							timestamp(),
							u64tostr (seg_list[last_seg_repacked].fileid, tmpbuf, 0),
							seg_list[last_seg_repacked].copyno,
							seg_list[last_seg_repacked].fsec,
							seg_list[last_seg_repacked].vid,
							seg_list[last_seg_repacked].side,
							seg_list[last_seg_repacked].fseq,
							seg_list[last_seg_repacked].blockid[0],
							seg_list[last_seg_repacked].blockid[1],
							seg_list[last_seg_repacked].blockid[2],
							seg_list[last_seg_repacked].blockid[3],
							seg_list[last_seg_repacked].compression,
							seg_list[last_seg_repacked].checksum_name,
							seg_list[last_seg_repacked].checksum,
							tl->vid, tl->side, fl->tape_fseq, fl->blockid[0], fl->blockid[1],
							fl->blockid[2], fl->blockid[3], compression_factor,
							fl->castorSegAttr.segmCksumAlgorithm, fl->castorSegAttr.segmCksum,
							sstrerror(serrno)
						);
					fflush (stdout);
					return(-1);
				}
			}
		}

		fileid = seg_list[last_seg_repacked].fileid;

		old_segattrs.copyno = seg_list[last_seg_repacked].copyno;
		old_segattrs.fsec = seg_list[last_seg_repacked].fsec;
		strcpy (old_segattrs.vid, seg_list[last_seg_repacked].vid);
		old_segattrs.side = seg_list[last_seg_repacked].side;
		old_segattrs.fseq = seg_list[last_seg_repacked].fseq;
		/* Old checksum is always there anyway - useless not to print it, if any */
		strcpy(old_segattrs.checksum_name,seg_list[last_seg_repacked].checksum_name);
		old_segattrs.checksum = seg_list[last_seg_repacked].checksum;

		new_segattrs.copyno = seg_list[last_seg_repacked].copyno;
		new_segattrs.fsec = seg_list[last_seg_repacked].fsec;
		new_segattrs.compression = compression_factor;
		strcpy (new_segattrs.vid, tl->vid);
		new_segattrs.side = tl->side;
		new_segattrs.fseq = fl->tape_fseq;
		memcpy (new_segattrs.blockid, fl->blockid, 4);

		if (use_checksum) {
			/* Propagate checksum that mover returned into the new segment structure */
			strncpy(new_segattrs.checksum_name, fl->castorSegAttr.segmCksumAlgorithm,CA_MAXCKSUMNAMELEN);
			new_segattrs.checksum_name[CA_MAXCKSUMNAMELEN] = '\0';
			new_segattrs.checksum = fl->castorSegAttr.segmCksum;
		} else {
			/* Make sure it is zero in the new segment structure */
			new_segattrs.checksum_name[0] = '\0';
		    new_segattrs.checksum = 0;
		}
		
		if (Cns_replaceseg (host, fileid, &old_segattrs, &new_segattrs) < 0) {
			fprintf (stderr, "Cns_replaceseg (%s %d %d) error: %s\n",
			    u64tostr (fileid, tmpbuf, 0), old_segattrs.copyno,
			    old_segattrs.fsec, sstrerror(serrno));
			if (serrno == EACCES) {
				usr_errflg++;
				return (-1);
			} else if (serrno != ENOENT) {
				sys_errflg++;
				return (-1);
			}
		}
		
		printf ("%s (%s,%d,%d) repacked from (%s,%d,%d,%02x%02x%02x%02x,%d,\"%s\",0x%lx) to (%s,%d,%d,%02x%02x%02x%02x,%d,\"%s\",0x%lx)\n",
				timestamp(),
				u64tostr (seg_list[last_seg_repacked].fileid, tmpbuf, 0),
				seg_list[last_seg_repacked].copyno,
				seg_list[last_seg_repacked].fsec,
				seg_list[last_seg_repacked].vid,
				seg_list[last_seg_repacked].side,
				seg_list[last_seg_repacked].fseq,
				seg_list[last_seg_repacked].blockid[0],
				seg_list[last_seg_repacked].blockid[1],
				seg_list[last_seg_repacked].blockid[2],
				seg_list[last_seg_repacked].blockid[3],
				seg_list[last_seg_repacked].compression,
				seg_list[last_seg_repacked].checksum_name,
				seg_list[last_seg_repacked].checksum,
				tl->vid, tl->side, fl->tape_fseq, fl->blockid[0], fl->blockid[1],
				fl->blockid[2], fl->blockid[3], compression_factor,
				fl->castorSegAttr.segmCksumAlgorithm, fl->castorSegAttr.segmCksum);
		fflush (stdout);

		if (! nodelete_flag) {
			printf ("%s Deleting %s.%d from disk\n", timestamp(),
			    seg_list[last_seg_repacked].vid,
			    seg_list[last_seg_repacked].fseq);
			fflush (stdout);
			sprintf(tmpbuf, "%d", seg_list[last_seg_repacked].fseq);
			(void) stage_clr_Tape ((u_signed64) STAGE_NOLINKCHECK,
			    NULL, seg_list[last_seg_repacked].vid, tmpbuf);
		}
		size_requested -= seg_list[last_seg_repacked].segsize;
		last_seg_repacked++;
		last_stcp_repacked++;
	}
	return (0);
}

void
repack_rtcplog(int level, CONST char *format, ...)
{
        va_list args;

	va_start (args, format);
	if (level != LOG_DEBUG)
		vprintf (format, args);
	va_end(args);
}

char *
timestamp()
{
	time_t current_time;
	static char timestr[16];
	struct tm *tm;

	(void) time (&current_time);
	tm = localtime (&current_time);
	(void) strftime(timestr, 16, "%b %e %H:%M:%S", tm);
	return (timestr);
}

write_segments(pool_name)
char *pool_name;
{
	int c;
	rtcpFileRequest_t filereq;
	u_signed64 estimated_free_space;
	file_list_t *fl = NULL;
	int Flags = 0;
	file_list_t *flt;
	int fseq;
	int i;
	char model[CA_MAXMODELLEN+1];
	u_signed64 requested_space;
	struct stgcat_entry *stcp;
	tape_list_t *tl = NULL;
	int save_serrno;

	rtcp_log = (void (*) _PROTO((int, CONST char *, ...))) &repack_rtcplog;

	/* build the tape list for rtcopy */

	if (rtcp_NewTapeList (&tl, NULL, WRITE_ENABLE) < 0) {
		fprintf (stderr, "repack error: %s\n", sstrerror(serrno));
		return (USERR);
	}

	/* get and lock an output tape */

	if (vmgr_gettape (pool_name, size_requested, NULL, tl->tapereq.vid,
	    tl->tapereq.vsn, tl->tapereq.dgn, tl->tapereq.density,
	    tl->tapereq.label, model, &tl->tapereq.side, &fseq,
	    &estimated_free_space) < 0) {
		fprintf (stderr, "vmgr_gettape error: %s\n",
		    sstrerror(serrno));
		return ((serrno == EACCES) ? USERR : SYERR);
	}
	out_vid = tl->tapereq.vid;
	out_side = tl->tapereq.side;
	printf ("%s Writing to tape %s\n", timestamp(), tl->tapereq.vid);
	fflush (stdout);

	/* build the file list for rtcopy */

	requested_space = 0;
	stcp = stcp_output + last_stcp_repacked;
	for (i = last_stcp_repacked; i < nstcp_output; i++) {
		requested_space += stcp->actual_size;
		if (requested_space > estimated_free_space) break;
		if (rtcp_NewFileList (&tl, &fl, WRITE_ENABLE) < 0) {
			fprintf (stderr, "repack error: %s\n", sstrerror(serrno));
			return (USERR);
		}
		strcpy (fl->filereq.file_path, stcp->ipath);
		strcpy (fl->filereq.recfm, "F");
		strcpy (fl->filereq.fid, stcp->u1.t.fid);
		fl->filereq.position_method = TPPOSIT_FSEQ;
		fl->filereq.tape_fseq = fseq + i - last_stcp_repacked;
		fl->filereq.disk_fseq = i + 1 - last_stcp_repacked;
		fl->filereq.check_fid = NEW_FILE;
		fl->filereq.concat = NOCONCAT;
		FILL_USE_CHECKSUM_VARIABLE();
		if (use_checksum) {
			/* We propage to the tpwrite the checksum we got from name server */
			strncpy(fl->filereq.castorSegAttr.segmCksumAlgorithm, seg_list[i].checksum_name,CA_MAXCKSUMNAMELEN);
			fl->filereq.castorSegAttr.segmCksumAlgorithm[CA_MAXCKSUMNAMELEN] = '\0';
			fl->filereq.castorSegAttr.segmCksum = seg_list[i].checksum;
		}
		stcp++;
	}
	last_rtcp_disk_fseq = fl->filereq.disk_fseq;
	rtcpc_ClientCallback = &repack_callback;

	rtcpc_kill_needed = 1;
	c = rtcpc (tl);
	rtcpc_kill_needed = 0;
	save_serrno = serrno;

	if (c) {
		switch (serrno) {
		case SECHECKSUM: /* Bad checksum */
			fprintf (stderr,"Repack failure, %s\n", sstrerror(serrno));
			break;
		case ETWLBL:	/* Wrong label type */
		case ETWVSN:	/* Wrong vsn */
		case ETHELD:	/* Volume held or disabled */
		case ETVUNKN:	/* Volume unknown to TMS */
		case ETOPAB:	/* Operator cancel */
			Flags = DISABLED;
			fprintf (stderr,"Setting %s status to DISABLED\n", tl->tapereq.vid);
			break;
		case ETABSENT:	/* Volume not in library/absent */
			Flags = EXPORTED;
			fprintf (stderr,"Setting %s status to EXPORTED\n", tl->tapereq.vid);
			break;
		case ETNXPD:	/* File not expired */
		case ETWPROT:	/* Cartridge physically write protected or tape read-only (TMS) */
			Flags = TAPE_RDONLY;
			fprintf (stderr,"Setting %s status to RDONLY\n", tl->tapereq.vid);
			break;
		case ETARCH:	/* Volume in inactive library */
			Flags = ARCHIVED;
			fprintf (stderr,"Setting %s status to ARCHIVED\n", tl->tapereq.vid);
			break;
		case ENOSPC:
			Flags = TAPE_RDONLY;
			fprintf (stderr,"Setting %s status to RDONLY (quite FULL)\n", tl->tapereq.vid);
			break;
		default:
			if (tl->tapereq.tprc &&
			   (tl->tapereq.err.errorcode == ETPARIT ||
			    tl->tapereq.err.errorcode == ETUNREC ||
			    tl->tapereq.err.errorcode == ETLBL)) {
				Flags = TAPE_RDONLY;
				fprintf (stderr,"Setting %s status to RDONLY\n",
				    tl->tapereq.vid);
				break;
			}
			fl = tl->file;
			do {
				if (fl->filereq.err.severity & RTCP_FAILED) {
					if (fl->filereq.err.errorcode == ETPARIT ||
					    fl->filereq.err.errorcode == ETUNREC ||
					    fl->filereq.err.errorcode == ETLBL) {
						Flags = TAPE_RDONLY;
						fprintf (stderr,"Setting %s status to RDONLY\n",
						    tl->tapereq.vid);
					}
					break;
				}
				fl = fl->next;
			} while (fl != tl->file);
			break;
		}
		c = (serrno == ENOSPC) ? 0 : usr_errflg ? USERR : SYERR;
		if (vmgr_updatetape (tl->tapereq.vid, tl->tapereq.side,
		    (u_signed64) 0, 0, 0, Flags) < 0) {
			fprintf (stderr, "vmgr_updatetape %s: %s\n", tl->tapereq.vid,
			    sstrerror(serrno));
			if (c == 0)
				c = (serrno == EACCES) ? USERR : SYERR;
		} else {
			serrno = save_serrno;
		}
	}
	out_vid = NULL;

	/* free rtcpc request */

	fl = tl->file;
	do {
		flt = fl;
		fl = fl->next;
		free (flt);
	} while (fl != tl->file);
	free (tl);
	
	if (c && serrno == SECHECKSUM) {
		c = CHECKSUMERR;
	}
	return (c);
}
